{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Annotating and clustering CARF-domain containing proteins\n",
    "\n",
    "__reference__:\n",
    "\n",
    "Makarova, Kira S, Albertas Timinskas, Yuri I Wolf, Ayal B Gussow, Virginijus Siksnys, Česlovas Venclovas, and Eugene V Koonin. “Evolutionary and Functional Classification of the CARF Domain Superfamily, Key Sensors in Prokaryotic Antivirus Defense.” Nucleic Acids Research 48, no. 16 (September 18, 2020): 8828–47. [https://doi.org/10.1093/nar/gkaa635](https://doi.org/10.1093/nar/gkaa635).\n",
    "\n",
    "## Workflow overview\n",
    "\n",
    "![Overview of CARF clustering](../media/CARF_workflow_web.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import pandas as pd\n",
    "from domainator.Bio.SeqIO.FastaIO import SimpleFastaParser\n",
    "from glob import glob\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "--2024-02-13 11:02:21--  https://ftp.ncbi.nlm.nih.gov/pub/makarova/Supplement/CARF/Supplementary_Data_File_1.tgz\n",
      "Resolving ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)... 130.14.250.10, 130.14.250.7, 2607:f220:41e:250::10, ...\n",
      "Connecting to ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)|130.14.250.10|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 1096439 (1.0M) [application/x-gzip]\n",
      "Saving to: ‘../resources/Supplementary_Data_File_1.tgz’\n",
      "\n",
      "     0K .......... .......... .......... .......... ..........  4% 1.55M 1s\n",
      "    50K .......... .......... .......... .......... ..........  9% 3.21M 0s\n",
      "   100K .......... .......... .......... .......... .......... 14%  200M 0s\n",
      "   150K .......... .......... .......... .......... .......... 18% 3.32M 0s\n",
      "   200K .......... .......... .......... .......... .......... 23%  195M 0s\n",
      "   250K .......... .......... .......... .......... .......... 28%  107M 0s\n",
      "   300K .......... .......... .......... .......... .......... 32%  133M 0s\n",
      "   350K .......... .......... .......... .......... .......... 37%  106M 0s\n",
      "   400K .......... .......... .......... .......... .......... 42% 3.54M 0s\n",
      "   450K .......... .......... .......... .......... .......... 46%  115M 0s\n",
      "   500K .......... .......... .......... .......... .......... 51%  119M 0s\n",
      "   550K .......... .......... .......... .......... .......... 56%  126M 0s\n",
      "   600K .......... .......... .......... .......... .......... 60% 96.1M 0s\n",
      "   650K .......... .......... .......... .......... .......... 65%  125M 0s\n",
      "   700K .......... .......... .......... .......... .......... 70%  119M 0s\n",
      "   750K .......... .......... .......... .......... .......... 74%  109M 0s\n",
      "   800K .......... .......... .......... .......... .......... 79% 4.11M 0s\n",
      "   850K .......... .......... .......... .......... .......... 84%  109M 0s\n",
      "   900K .......... .......... .......... .......... .......... 88%  128M 0s\n",
      "   950K .......... .......... .......... .......... .......... 93%  113M 0s\n",
      "  1000K .......... .......... .......... .......... .......... 98%  114M 0s\n",
      "  1050K .......... ..........                                 100%  134M=0.09s\n",
      "\n",
      "2024-02-13 11:02:22 (11.2 MB/s) - ‘../resources/Supplementary_Data_File_1.tgz’ saved [1096439/1096439]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "# Download Pfam-A.hmm\n",
    "\n",
    "mkdir -p ../resources\n",
    "\n",
    "if [ ! -f ../resources/Pfam-A.hmm ]; then\n",
    "    wget -O ../resources/Pfam-A.hmm.gz https://ftp.ebi.ac.uk/pub/databases/Pfam/releases/Pfam36.0/Pfam-A.hmm.gz --no-check-certificate\n",
    "    gunzip ../resources/Pfam-A.hmm.gz\n",
    "fi\n",
    "\n",
    "# Downlad supplementary data for https://doi.org/10.1093/nar/gkaa635.\n",
    "\n",
    "if [ ! -d ../resources/CARF_SAVED ]; then\n",
    "    wget https://ftp.ncbi.nlm.nih.gov/pub/makarova/Supplement/CARF/Supplementary_Data_File_1.tgz -O ../resources/Supplementary_Data_File_1.tgz --no-check-certificate\n",
    "    tar x -C ../resources -f ../resources/Supplementary_Data_File_1.tgz\n",
    "fi\n",
    "\n",
    "# Download CARF.fasta\n",
    "if [ ! -f ../resources/CARF_full.fasta ]; then\n",
    "    wget https://ftp.ncbi.nlm.nih.gov/pub/makarova/Supplement/CARF/CARF.fasta.gz -O ../resources/CARF.fasta.gz --no-check-certificate\n",
    "    gunzip ../resources/CARF.fasta.gz\n",
    "    mv ../resources/CARF.fasta ../resources/CARF_full.fasta\n",
    "fi\n",
    "\n",
    "# download Makarova et al. 2020 table S1. Already in github repository as ..resources/CARF_metadata.tsv\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove consensus sequences from CARF\n",
    "# consensus is the first sequence in each file\n",
    "\n",
    "!mkdir -p CARF_SAVED_no_consensus\n",
    "for f in glob(\"../resources/CARF_SAVED/*.FASTA\"):\n",
    "    f = Path(f)\n",
    "    f_basename = f.stem\n",
    "    #CARF_name = f_basename.split(\".\")[1]\n",
    "    with open(f, \"r\") as handle:\n",
    "        with open(f\"CARF_SAVED_no_consensus/{f_basename}.fasta\", \"w\") as out:\n",
    "            for i, (title, seq) in enumerate(SimpleFastaParser(handle)):\n",
    "                if i != 0:\n",
    "                    out.write(f\">{f_basename}_{title}\\n{seq}\\n\")\n",
    "    subprocess.run([\"hmmer_build.py\", \"-i\", f\"CARF_SAVED_no_consensus/{f_basename}.fasta\", \"-o\", f\"CARF_SAVED_no_consensus/{f_basename}.hmm\", \"--name\", f\"{f_basename}\", \"--acc\", f\"{f_basename}\", \"--desc\", f\"{f_basename}\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove consensus sequences from CARF\n",
    "# consensus is the first sequence in each file\n",
    "!mkdir -p effector_no_consensus\n",
    "for f in glob(\"../resources/effector_domain/*.FASTA\"):\n",
    "    f = Path(f)\n",
    "    f_basename = f.stem\n",
    "    with open(f, \"r\") as handle:\n",
    "        with open(f\"effector_no_consensus/{f_basename}.fasta\", \"w\") as out:\n",
    "            for i, (title, seq) in enumerate(SimpleFastaParser(handle)):\n",
    "                if i != 0:\n",
    "                    out.write(f\">{title}\\n{seq}\\n\")\n",
    "    subprocess.run([\"hmmer_build.py\", \"-i\", f\"effector_no_consensus/{f_basename}.fasta\", \"-o\", f\"effector_no_consensus/{f_basename}.hmm\", \"--name\", f\"{f_basename}\", \"--acc\", f\"{f_basename}\", \"--desc\", f\"{f_basename}\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# concatenate hmm files\n",
    "!rm -f CARF_SAVED_no_consensus/CARF.hmm\n",
    "!cat CARF_SAVED_no_consensus/*.hmm > CARF_SAVED_no_consensus/CARF.hmm\n",
    "#!ln -s CARF_SAVED_no_consensus/CARF_SAVED_domains.hmm CARF_SAVED_no_consensus/CARF_SAVED_copy.hmm # make a symlink so that we can search twice with the same database in different ways with the same \n",
    "#!cp CARF_SAVED_no_consensus/CARF_SAVED_domains.hmm CARF_SAVED_no_consensus/CARF_SAVED_copy.hmm # make a symlink so that we can search twice with the same database in different ways with the same \n",
    "\n",
    "!rm -f effector_no_consensus/effectors.hmm\n",
    "!cat effector_no_consensus/*.hmm > effector_no_consensus/effectors.hmm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = pd.read_csv(\"../resources/CARF_metadata.tsv\", sep=\"\\t\")\n",
    "metadata[\"Final group assignments\"].fillna(\"\", inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Local ID</th>\n",
       "      <th>Other Protein ID</th>\n",
       "      <th>Profile superfamily</th>\n",
       "      <th>Profile ID</th>\n",
       "      <th>E-value</th>\n",
       "      <th>Identity cluster</th>\n",
       "      <th>Order of leaves in full sequence tree</th>\n",
       "      <th>Superclusters based on CARF/SAVED domain tree</th>\n",
       "      <th>Final group assignments</th>\n",
       "      <th>Genome Name</th>\n",
       "      <th>Protein Length</th>\n",
       "      <th>CDD asignments, automatic with E-value 0.0001 (CARF only)</th>\n",
       "      <th>Domain organization</th>\n",
       "      <th>TMHMM result</th>\n",
       "      <th>Current genbank annotation</th>\n",
       "      <th>Potential Ring Nucleases</th>\n",
       "      <th>RN group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MK_CARF_00011497</td>\n",
       "      <td>WP_014788072.1</td>\n",
       "      <td>CARF</td>\n",
       "      <td>CARF.89.FASTA</td>\n",
       "      <td>3.180000e-57</td>\n",
       "      <td>3606</td>\n",
       "      <td>2524</td>\n",
       "      <td>CL1_CARF</td>\n",
       "      <td>CARF_m1</td>\n",
       "      <td>Thermococcus cleftensis</td>\n",
       "      <td>160</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CARF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>RN(1)</td>\n",
       "      <td>RN_2_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MK_CARF_00101059</td>\n",
       "      <td>YP_006424225.1</td>\n",
       "      <td>CARF</td>\n",
       "      <td>CARF.89.FASTA</td>\n",
       "      <td>3.180000e-57</td>\n",
       "      <td>3606</td>\n",
       "      <td>2525</td>\n",
       "      <td>CL1_CARF</td>\n",
       "      <td>CARF_m1</td>\n",
       "      <td>Thermococcus cleftensis</td>\n",
       "      <td>160</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CARF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein CL1_0219</td>\n",
       "      <td>RN(1)</td>\n",
       "      <td>RN_2_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MK_CARF_00011498</td>\n",
       "      <td>WP_088864412.1</td>\n",
       "      <td>CARF</td>\n",
       "      <td>CARF.89.FASTA</td>\n",
       "      <td>3.180000e-57</td>\n",
       "      <td>1115</td>\n",
       "      <td>2526</td>\n",
       "      <td>CL1_CARF</td>\n",
       "      <td>CARF_m1</td>\n",
       "      <td>Thermococcus barossii</td>\n",
       "      <td>160</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CARF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>RN(1)</td>\n",
       "      <td>RN_2_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MK_CARF_00011442</td>\n",
       "      <td>WP_088180185.1</td>\n",
       "      <td>CARF</td>\n",
       "      <td>CARF.89.FASTA</td>\n",
       "      <td>3.890000e-58</td>\n",
       "      <td>2701</td>\n",
       "      <td>2527</td>\n",
       "      <td>CL1_CARF</td>\n",
       "      <td>CARF_m1</td>\n",
       "      <td>Thermococcus sp. 5-4</td>\n",
       "      <td>160</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CARF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>RN(1)</td>\n",
       "      <td>RN_2_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MK_CARF_00011410</td>\n",
       "      <td>WP_088866426.1</td>\n",
       "      <td>CARF</td>\n",
       "      <td>CARF.89.FASTA</td>\n",
       "      <td>6.760000e-59</td>\n",
       "      <td>2988</td>\n",
       "      <td>2528</td>\n",
       "      <td>CL1_CARF</td>\n",
       "      <td>CARF_m1</td>\n",
       "      <td>Thermococcus radiotolerans</td>\n",
       "      <td>160</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CARF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>RN(1)</td>\n",
       "      <td>RN_2_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7138</th>\n",
       "      <td>MK_CARF_00012566</td>\n",
       "      <td>WP_051667291.1</td>\n",
       "      <td>SAVED</td>\n",
       "      <td>SAVED.104.FASTA</td>\n",
       "      <td>5.840000e-39</td>\n",
       "      <td>3587</td>\n",
       "      <td>5367</td>\n",
       "      <td>cl1_SAVED</td>\n",
       "      <td></td>\n",
       "      <td>Bacillus lehensis</td>\n",
       "      <td>492</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAVED_ABhydrolase</td>\n",
       "      <td>NaN</td>\n",
       "      <td>alpha/beta hydrolase</td>\n",
       "      <td>Non-RN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7139</th>\n",
       "      <td>MK_CARF_ADD00000372</td>\n",
       "      <td>PYQ55858.1</td>\n",
       "      <td>SAVED</td>\n",
       "      <td>SAVED.98.FASTA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>135a</td>\n",
       "      <td>6210</td>\n",
       "      <td>cl1_SAVED</td>\n",
       "      <td></td>\n",
       "      <td>Acidobacteria bacterium</td>\n",
       "      <td>555</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAVED_HATPase_Alba</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein DMF53_27210</td>\n",
       "      <td>Non-RN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7140</th>\n",
       "      <td>MK_CARF_ADD00000373</td>\n",
       "      <td>WP_052549029.1</td>\n",
       "      <td>SAVED</td>\n",
       "      <td>SAVED.98.FASTA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>139a</td>\n",
       "      <td>6211</td>\n",
       "      <td>cl1_SAVED</td>\n",
       "      <td></td>\n",
       "      <td>Enhygromyxa salina</td>\n",
       "      <td>551</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAVED_HATPase_Alba</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAVED domain-containing protein</td>\n",
       "      <td>Non-RN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7141</th>\n",
       "      <td>MK_CARF_00013068</td>\n",
       "      <td>WP_021722632.1</td>\n",
       "      <td>SAVED</td>\n",
       "      <td>SAVED.104.FASTA</td>\n",
       "      <td>4.040000e-20</td>\n",
       "      <td>3740</td>\n",
       "      <td>6212</td>\n",
       "      <td>cl1_SAVED</td>\n",
       "      <td></td>\n",
       "      <td>Lactococcus lactis</td>\n",
       "      <td>413</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAVED</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>Non-RN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7142</th>\n",
       "      <td>MK_CARF_00012153</td>\n",
       "      <td>WP_080810306.1</td>\n",
       "      <td>SAVED</td>\n",
       "      <td>SAVED.104.FASTA</td>\n",
       "      <td>5.590000e-40</td>\n",
       "      <td>777</td>\n",
       "      <td>7143</td>\n",
       "      <td>cl1_SAVED</td>\n",
       "      <td></td>\n",
       "      <td>Halomicronema hongdechloris</td>\n",
       "      <td>544</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAVED</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>Non-RN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7143 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Local ID Other Protein ID Profile superfamily  \\\n",
       "0        MK_CARF_00011497   WP_014788072.1                CARF   \n",
       "1        MK_CARF_00101059   YP_006424225.1                CARF   \n",
       "2        MK_CARF_00011498   WP_088864412.1                CARF   \n",
       "3        MK_CARF_00011442   WP_088180185.1                CARF   \n",
       "4        MK_CARF_00011410   WP_088866426.1                CARF   \n",
       "...                   ...              ...                 ...   \n",
       "7138     MK_CARF_00012566   WP_051667291.1               SAVED   \n",
       "7139  MK_CARF_ADD00000372       PYQ55858.1               SAVED   \n",
       "7140  MK_CARF_ADD00000373   WP_052549029.1               SAVED   \n",
       "7141     MK_CARF_00013068   WP_021722632.1               SAVED   \n",
       "7142     MK_CARF_00012153   WP_080810306.1               SAVED   \n",
       "\n",
       "           Profile ID       E-value Identity cluster  \\\n",
       "0       CARF.89.FASTA  3.180000e-57             3606   \n",
       "1       CARF.89.FASTA  3.180000e-57             3606   \n",
       "2       CARF.89.FASTA  3.180000e-57             1115   \n",
       "3       CARF.89.FASTA  3.890000e-58             2701   \n",
       "4       CARF.89.FASTA  6.760000e-59             2988   \n",
       "...               ...           ...              ...   \n",
       "7138  SAVED.104.FASTA  5.840000e-39             3587   \n",
       "7139   SAVED.98.FASTA           NaN             135a   \n",
       "7140   SAVED.98.FASTA           NaN             139a   \n",
       "7141  SAVED.104.FASTA  4.040000e-20             3740   \n",
       "7142  SAVED.104.FASTA  5.590000e-40              777   \n",
       "\n",
       "      Order of leaves in full sequence tree  \\\n",
       "0                                      2524   \n",
       "1                                      2525   \n",
       "2                                      2526   \n",
       "3                                      2527   \n",
       "4                                      2528   \n",
       "...                                     ...   \n",
       "7138                                   5367   \n",
       "7139                                   6210   \n",
       "7140                                   6211   \n",
       "7141                                   6212   \n",
       "7142                                   7143   \n",
       "\n",
       "     Superclusters based on CARF/SAVED domain tree Final group assignments  \\\n",
       "0                                         CL1_CARF                 CARF_m1   \n",
       "1                                         CL1_CARF                 CARF_m1   \n",
       "2                                         CL1_CARF                 CARF_m1   \n",
       "3                                         CL1_CARF                 CARF_m1   \n",
       "4                                         CL1_CARF                 CARF_m1   \n",
       "...                                            ...                     ...   \n",
       "7138                                     cl1_SAVED                           \n",
       "7139                                     cl1_SAVED                           \n",
       "7140                                     cl1_SAVED                           \n",
       "7141                                     cl1_SAVED                           \n",
       "7142                                     cl1_SAVED                           \n",
       "\n",
       "                      Genome Name  Protein Length  \\\n",
       "0         Thermococcus cleftensis             160   \n",
       "1         Thermococcus cleftensis             160   \n",
       "2           Thermococcus barossii             160   \n",
       "3            Thermococcus sp. 5-4             160   \n",
       "4      Thermococcus radiotolerans             160   \n",
       "...                           ...             ...   \n",
       "7138            Bacillus lehensis             492   \n",
       "7139      Acidobacteria bacterium             555   \n",
       "7140           Enhygromyxa salina             551   \n",
       "7141           Lactococcus lactis             413   \n",
       "7142  Halomicronema hongdechloris             544   \n",
       "\n",
       "     CDD asignments, automatic with E-value 0.0001 (CARF only)  \\\n",
       "0                                                   NaN          \n",
       "1                                                   NaN          \n",
       "2                                                   NaN          \n",
       "3                                                   NaN          \n",
       "4                                                   NaN          \n",
       "...                                                 ...          \n",
       "7138                                                NaN          \n",
       "7139                                                NaN          \n",
       "7140                                                NaN          \n",
       "7141                                                NaN          \n",
       "7142                                                NaN          \n",
       "\n",
       "     Domain organization  TMHMM result        Current genbank annotation  \\\n",
       "0                   CARF           NaN              hypothetical protein   \n",
       "1                   CARF           NaN     hypothetical protein CL1_0219   \n",
       "2                   CARF           NaN              hypothetical protein   \n",
       "3                   CARF           NaN              hypothetical protein   \n",
       "4                   CARF           NaN              hypothetical protein   \n",
       "...                  ...           ...                               ...   \n",
       "7138   SAVED_ABhydrolase           NaN              alpha/beta hydrolase   \n",
       "7139  SAVED_HATPase_Alba           NaN  hypothetical protein DMF53_27210   \n",
       "7140  SAVED_HATPase_Alba           NaN   SAVED domain-containing protein   \n",
       "7141               SAVED           NaN              hypothetical protein   \n",
       "7142               SAVED           NaN              hypothetical protein   \n",
       "\n",
       "     Potential Ring Nucleases RN group  \n",
       "0                       RN(1)   RN_2_5  \n",
       "1                       RN(1)   RN_2_5  \n",
       "2                       RN(1)   RN_2_5  \n",
       "3                       RN(1)   RN_2_5  \n",
       "4                       RN(1)   RN_2_5  \n",
       "...                       ...      ...  \n",
       "7138                   Non-RN      NaN  \n",
       "7139                   Non-RN      NaN  \n",
       "7140                   Non-RN      NaN  \n",
       "7141                   Non-RN      NaN  \n",
       "7142                   Non-RN      NaN  \n",
       "\n",
       "[7143 rows x 17 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1844"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Select CARF and only CARF, exclude proteins with previously annotated fusions\n",
    "CARF_metadata = metadata[(metadata[\"Profile superfamily\"] == \"CARF\") & ((metadata[\"Domain organization\"] == \"CARF\"))]\n",
    "len(CARF_metadata[\"Local ID\"].tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get just the CARF sequences, not the SAVED sequences, and not the RtcR CARFs.\n",
    "subprocess.run([\"select_by_contig.py\", \"-i\", \"../resources/CARF_full.fasta\", \"-o\", \"no_effector.gb\", \"--contigs\"] + CARF_metadata[\"Local ID\"].tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args=['domainate.py', '-i', 'no_effector.gb', '-r', 'CARF_SAVED_no_consensus/CARF.hmm', '-o', 'CARF_loc.gb', '-e', '1e-3'], returncode=0)"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# annotate sequences with the best CARF domain\n",
    "subprocess.run([\"domainate.py\", \"-i\", \"no_effector.gb\", \"-r\", \"CARF_SAVED_no_consensus/CARF.hmm\", \"-o\", \"CARF_loc.gb\", \"-e\", \"1e-3\"])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract the longest regions ( >= 50 aa), not containing the best CARF domain.\n",
    "! extract_unannotated.py -i CARF_loc.gb -o unannotated.gb --largest --lb 50 --keep_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add pfam and effector domain annotations\n",
    "! domainate.py -i unannotated.gb -r effector_no_consensus/effectors.hmm -e 0.001 --max_overlap 0.6 -o tmp.gb\n",
    "\n",
    "! domainate.py -i tmp.gb -r ../resources/Pfam-A.hmm -e 0.001 --max_overlap 0.6 -o new_annots.gb\n",
    "\n",
    "! rm -f tmp.gb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate pairwise sequence similarity scores between all proteins\n",
    "! seq_dist.py -i new_annots.gb  -r new_annots.gb  --sparse scores.hdf5 --mode efi_score "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Tabular report where each row is a single contig\n",
    "! enum_report.py -i new_annots.gb --domains --domain_descriptions -o annotations.tsv --length --sequence "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Contig Stats\n",
      "\n",
      "contigs: 1524\n",
      "CDSs: 0\n",
      "CDSs per 10 kb:  0.00\n",
      "\n",
      "    --------------------------------------------------------------------------------------------------------\n",
      "    |                                            Contig Lengths                                            |\n",
      "    --------------------------------------------------------------------------------------------------------\n",
      "\n",
      " 248|    o                                               \n",
      " 235|    o                                               \n",
      " 222|    o                                               \n",
      " 209|    o                                               \n",
      " 196|    o                                               \n",
      " 183|    o                                               \n",
      " 170|   oo                                               \n",
      " 157|   oo                                               \n",
      " 144|   ooo                                              \n",
      " 131|   ooo                                              \n",
      " 118|   ooo                                              \n",
      " 105|   ooo   o                                          \n",
      "  92|   ooo   o   o                                      \n",
      "  79|   ooo   o   o oo                                   \n",
      "  66|  oooo o oo  o oo                                   \n",
      "  53|  ooooooooo  o oo                                   \n",
      "  40|  ooooooooo  ooooo                                  \n",
      "  27|  ooooooooo  ooooo                                  \n",
      "  14|  ooooooooooooooooo                                 \n",
      "   1| oooooooooooooooooooooooooo o oo o  ooooo o o     oo\n",
      "     ---------------------------------------------------\n",
      "     5 6 8 1 1 1 1 1 1 2 2 2 2 2 2 3 3 3 3 3 3 4 4 4 4 4 \n",
      "     1 8 5 0 2 3 5 7 9 0 2 4 6 7 9 1 3 4 6 8 9 1 3 5 6 8 \n",
      "       . . 3 0 8 5 3 0 7 5 2 0 7 5 2 0 7 4 2 9 7 4 2 9 7 \n",
      "       4 8 . . . . . . . . . . . . . . . . . . . . . .   \n",
      "       4 8 3 7 2 6 0 5 9 4 8 2 7 1 6 0 4 9 3 8 2 6 1 5   \n",
      "           2 6   4 8 2 6   4 8 2 6   4 8 2 6   4 8 2 6   \n",
      "\n",
      "Domain Stats\n",
      "        \n",
      "        Domain Frequency\n",
      "\n",
      "domain                database   count description                                                        avg score\n",
      "              DUF6293    Pfam-A 142                                 Family of unknown function (DUF6293)  47       \n",
      "                HTH_5    Pfam-A  96                            Bacterial regulatory protein, arsR family  30       \n",
      "               MarR_2    Pfam-A  71                                                          MarR family  30       \n",
      "               HTH_24    Pfam-A  57                                  Winged helix-turn-helix DNA-binding  29       \n",
      "                 TrmB    Pfam-A  54                        Sugar-specific transcriptional regulator TrmB  26       \n",
      "                 GerE    Pfam-A  53                           Bacterial regulatory proteins, luxR family  33       \n",
      "              DUF4364    Pfam-A  49                                 Domain of unknown function (DUF4364)  16       \n",
      "             HTH_IclR    Pfam-A  49                                         IclR helix-turn-helix domain  32       \n",
      "               HTH_38    Pfam-A  41                                              Helix-turn-helix domain  32       \n",
      "               HTH_20    Pfam-A  37                                              Helix-turn-helix domain  28       \n",
      "        CARF7_DUF2103 effectors  30                                                        CARF7_DUF2103  28       \n",
      "              HTH_Mga    Pfam-A  23           M protein trans-acting positive regulator (MGA) HTH domain  25       \n",
      "               HTH_34    Pfam-A  22                                      Winged helix DNA-binding domain  20       \n",
      "      Penicillinase_R    Pfam-A  21                                              Penicillinase repressor  28       \n",
      "                 MarR    Pfam-A  21                                                          MarR family  30       \n",
      "            HTH_Crp_2    Pfam-A  16                                     Crp-like helix-turn-helix domain  22       \n",
      "               HTH_45    Pfam-A  14                                              Winged helix-turn-helix  28       \n",
      "                 TnsD    Pfam-A  11                                     Tn7-like transposition protein D  15       \n",
      "               HTH_12    Pfam-A  10                                   Ribonuclease R winged-helix domain  16       \n",
      "            Cas_DxTHG    Pfam-A   9                                 CRISPR-associated (Cas) DxTHG family  47       \n",
      "         CARF2_b_HEPN effectors   9                                                         CARF2_b_HEPN  73       \n",
      "               HTH_27    Pfam-A   9                                      Winged helix DNA-binding domain  23       \n",
      "         CARF2_h_HEPN effectors   7                                                         CARF2_h_HEPN 123       \n",
      "              DUF6429    Pfam-A   7                                 Domain of unknown function (DUF6429)  17       \n",
      "          Sigma70_ECF    Pfam-A   7                                                     ECF sigma factor  22       \n",
      "                MCM_C    Pfam-A   6                MCM protein C-terminal winged helix-turn-helix domain  24       \n",
      "                HTH_1    Pfam-A   6           Bacterial regulatory helix-turn-helix protein, lysR family  21       \n",
      "         Sigma70_r4_2    Pfam-A   6                                                   Sigma-70, region 4  28       \n",
      "              DUF3071    Pfam-A   5                                Protein of unknown function (DUF3071)  18       \n",
      "           Cas_NE0113    Pfam-A   4                        CRISPR-associated protein NE0113 (Cas_NE0113)  32       \n",
      "               HTH_29    Pfam-A   4                                              Winged helix-turn helix  28       \n",
      "         Cas_Cas02710    Pfam-A   4                             CRISPR-associated protein (Cas_Cas02710)  40       \n",
      "                Mrr_N    Pfam-A   4                                                Mrr N-terminal domain  20       \n",
      "               HTH_23    Pfam-A   4                                              Homeodomain-like domain  26       \n",
      "         HrcA_DNA-bdg    Pfam-A   4    Winged helix-turn-helix transcription repressor, HrcA DNA-binding  25       \n",
      "            DASH_Dad2    Pfam-A   3                                            DASH complex subunit Dad2  16       \n",
      "              UPF0122    Pfam-A   3                   Putative helix-turn-helix protein, YlxM / p13 like  22       \n",
      "               HTH_11    Pfam-A   3                                                           HTH domain  24       \n",
      "                 VbhA    Pfam-A   3                                                       Antitoxin VbhA  15       \n",
      "        HTH_AsnC-type    Pfam-A   3                                    AsnC-type helix-turn-helix domain  16       \n",
      "      Phage_int_SAM_5    Pfam-A   3                                      Phage integrase SAM-like domain  16       \n",
      "        LexA_DNA_bind    Pfam-A   3                                              LexA DNA binding domain  29       \n",
      "         Topoisom_bac    Pfam-A   3                                                    DNA topoisomerase  16       \n",
      "               HupH_C    Pfam-A   3     HupH hydrogenase expression protein, C-terminal conserved region  20       \n",
      "                 PaaX    Pfam-A   3                                                    PaaX-like protein  21       \n",
      "       Arc_capsid_dom    Pfam-A   2    Activity-regulated cytoskeleton-associated protein, capsid domain  17       \n",
      "            RemA-like    Pfam-A   2                       Extracellular matrix regulatory protein A-like  18       \n",
      "         Replic_Relax    Pfam-A   2                                               Replication-relaxation  17       \n",
      "         CARF2_g_HEPN effectors   2                                                         CARF2_g_HEPN  68       \n",
      "                 PadR    Pfam-A   2                           Transcriptional regulator PadR-like family  20       \n",
      "              DUF5603    Pfam-A   2                                 Domain of unknown function (DUF5603)  16       \n",
      "            Abi_alpha    Pfam-A   2                                             Abortive infection alpha  17       \n",
      "                  NAP    Pfam-A   2                                    Nucleosome assembly protein (NAP)  15       \n",
      "              DUF3363    Pfam-A   2                                Protein of unknown function (DUF3363)  18       \n",
      "              AbiEi_4    Pfam-A   2                           Transcriptional regulator, AbiEi antitoxin  26       \n",
      "         CARF9_HEPN_6 effectors   2                                                         CARF9_HEPN_6  70       \n",
      "               RuvB_C    Pfam-A   2                                  RuvB C-terminal winged helix domain  17       \n",
      "              PFOR_II    Pfam-A   2                    Pyruvate:ferredoxin oxidoreductase core domain II  16       \n",
      "                 HxlR    Pfam-A   2                                           HxlR-like helix-turn-helix  22       \n",
      "            M20_dimer    Pfam-A   2                                        Peptidase dimerisation domain  25       \n",
      "          CARF_m1_PIN effectors   2                                                          CARF_m1_PIN  52       \n",
      "                SERTA    Pfam-A   2                                                          SERTA motif  17       \n",
      "              DUF5645    Pfam-A   2                                 Domain of unknown function (DUF5645)  15       \n",
      "              DUF6037    Pfam-A   1                                 Family of unknown function (DUF6037)  18       \n",
      "      FeoB_associated    Pfam-A   1                            FeoB-associated Cys-rich membrane protein  19       \n",
      "    CARF_m9_CorA_like effectors   1                                                    CARF_m9_CorA_like  73       \n",
      "               HTH_28    Pfam-A   1                                              Helix-turn-helix domain  25       \n",
      "            p25-alpha    Pfam-A   1                                                            p25-alpha  15       \n",
      "CARF7_RelE_final.x1.4 effectors   1                                                CARF7_RelE_final.x1.4  64       \n",
      "         mRNA_decap_C    Pfam-A   1                                     mRNA-decapping enzyme C-terminus  16       \n",
      "              DUF6499    Pfam-A   1                                 Family of unknown function (DUF6499)  15       \n",
      "               HemN_C    Pfam-A   1                                               HemN C-terminal domain  20       \n",
      "          RelA_AH_RIS    Pfam-A   1                                        RelA/SpoT, AH and RIS domains  15       \n",
      "         CARF9_HEPN_7 effectors   1                                                         CARF9_HEPN_7  21       \n",
      "       Fe_dep_repress    Pfam-A   1              Iron dependent repressor, N-terminal DNA binding domain  23       \n",
      "               PfUIS3    Pfam-A   1                          Plasmodium falciparum UIS3 membrane protein  16       \n",
      "              DUF1670    Pfam-A   1                                Protein of unknown function (DUF1670)  17       \n",
      "       STT3-PglB_core    Pfam-A   1                                           STT3/PglB/AglB core domain  16       \n",
      "              SLATT_1    Pfam-A   1                     SMODS and SLOG-associating 2TM effector domain 1  16       \n",
      "         CARF2_a_HEPN effectors   1                                                         CARF2_a_HEPN  20       \n",
      "                  NID    Pfam-A   1                                              Nmi/IFP 35 domain (NID)  16       \n",
      "              DUF3964    Pfam-A   1                                Protein of unknown function (DUF3964)  16       \n",
      "              BLOC1S3    Pfam-A   1        Biogenesis of lysosome-related organelles complex 1 subunit 3  15       \n",
      "                 Rrf2    Pfam-A   1                             Iron-dependent Transcriptional regulator  30       \n",
      "                 GntR    Pfam-A   1                           Bacterial regulatory proteins, gntR family  17       \n",
      "          CARF1_HEPN3 effectors   1                                                          CARF1_HEPN3  19       \n",
      "         CARF_m6_Unk1 effectors   1                                                         CARF_m6_Unk1  15       \n",
      "              DUF3116    Pfam-A   1                                Protein of unknown function (DUF3116)  15       \n",
      "                 RP-C    Pfam-A   1                              Replication protein C N-terminal domain  16       \n",
      "              DUF5528    Pfam-A   1                                 Family of unknown function (DUF5528)  15       \n",
      "              DUF6178    Pfam-A   1                                 Family of unknown function (DUF6178)  14       \n",
      "              DUF2089    Pfam-A   1                                Protein of unknown function (DUF2089)  16       \n",
      "              DUF1376    Pfam-A   1                                Protein of unknown function (DUF1376)  17       \n",
      "      AdenylateSensor    Pfam-A   1                         Adenylate sensor of SNF1-like protein kinase  17       \n",
      "             CwfJ_C_2    Pfam-A   1                                 Protein similar to CwfJ C-terminus 2  16       \n",
      "                 OSCP    Pfam-A   1                                    ATP synthase delta (OSCP) subunit  17       \n",
      "              DUF4518    Pfam-A   1                                 Domain of unknown function (DUF4518)  15       \n",
      "                 YjbF    Pfam-A   1                Group 4 capsule polysaccharide lipoprotein gfcB, YjbF  16       \n",
      "                 RecX    Pfam-A   1                                                          RecX family  16       \n",
      "          Sigma_reg_N    Pfam-A   1                                    Sigma factor regulator N-terminal  17       \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Summary of contig lengths and domain content\n",
    "! summary_report.py -i new_annots.gb --html summary.html  -o /dev/stdout"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Matrix Report\n",
      "Total values: 2322576\n",
      "Non-zero values: 78500\n",
      "Mean: 25.97977482447599\n",
      "Median: 16.35729514783766\n",
      "Min: 4.878998615944094\n",
      "Max: 270.06738811011354\n",
      "        \n",
      "\n",
      "      ------------------------------------------------------------------------------------------------------\n",
      "      |                                           Matrix Values                                            |\n",
      "      ------------------------------------------------------------------------------------------------------\n",
      "\n",
      " 24957|  o                                                \n",
      " 23644|  o                                                \n",
      " 22330|  o                                                \n",
      " 21017|  o                                                \n",
      " 19703|  o                                                \n",
      " 18390|  o                                                \n",
      " 17076|  o                                                \n",
      " 15763|  o                                                \n",
      " 14449|  o                                                \n",
      " 13136|  o                                                \n",
      " 11822|  o                                                \n",
      " 10509|  oo                                               \n",
      "  9195|  oo                                               \n",
      "  7882|  ooo       o                                      \n",
      "  6568|  ooo       o                                      \n",
      "  5255|  oooo      o                                      \n",
      "  3941|  oooo      o                                      \n",
      "  2628|  ooooo     oo                                     \n",
      "  1314|  ooooooo  oooo                                    \n",
      "     1| oooooooooooooooooooooooooooo oo  oooooooo   ooo  o\n",
      "       --------------------------------------------------\n",
      "       4 1 2 3 4 5 6 7 8 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 \n",
      "       . 5 6 6 7 7 8 9 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 \n",
      "       8 . . . . . . . . 0 0 1 2 2 3 3 4 5 5 6 7 7 8 8 9 \n",
      "       7 4 0 7 3 9 5 1 7 . . . . . . . . . . . . . . . . \n",
      "       8 8 9 0 0 1 2 3 3 3 9 5 1 7 3 9 5 2 8 4 0 6 2 8 4 \n",
      "       9 6 4 1 9 6 4 1 9 4 5 6 6 7 8 9 9 0 1 2 2 3 4 5 5 \n",
      "       9 5 0 6 1 6 2 7 2 6 4 1 9 6 4 2 9 7 4 2 9 7 4 2 9 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Summary of similarity matrix\n",
    "! matrix_report.py -i scores.hdf5 -o /dev/stdout"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate an SSN, colored by CARF group.\n",
    "# After the SSN is built, it can be visualized with Cytoscape. Use the Layout->Prefuse Force Directed Layout to visualize the network.\n",
    "! build_ssn.py -i scores.hdf5 --lb 5 --xgmml network.xgmml --metadata annotations.tsv ../resources/CARF_metadata.tsv --color_by \"Final group assignments\" --cluster_tsv clusters.tsv --cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### The rest of these steps are not depicted in the flow chart figure\n",
    "\n",
    "\n",
    "# make MSAs and compare clusters to Pfam\n",
    "clusters_table = pd.read_table(\"clusters.tsv\", names=[\"Name\", \"cluster\"])\n",
    "for i in range(0, 12):\n",
    "    idx = i + 1\n",
    "    clusters_table[clusters_table[\"cluster\"] == idx][\"Name\"].to_csv(f\"cluster_{idx}.tsv\", sep=\"\\t\", index=False, header=False)\n",
    "    subprocess.run([\"select_by_contig.py\", \"-i\", \"new_annots.gb\", \"-o\", f\"effector_cluster_{idx}.gb\", \"--contigs_file\", f\"cluster_{idx}.tsv\"])\n",
    "    subprocess.run([\"genbank_to_fasta.py\", \"-i\", f\"effector_cluster_{idx}.gb\", \"-o\", f\"effector_cluster_{idx}.fasta\"])\n",
    "    subprocess.run([\"mafft\", \"--maxiterate\", \"1000\", \"--globalpair\", f\"effector_cluster_{idx}.fasta\"], stdout=open(f\"effector_cluster_{idx}.mafft.fasta\", \"w\"))\n",
    "for i in range(0, 12):\n",
    "    idx = i + 1\n",
    "    subprocess.run([\"hmmer_build.py\", \"-i\", f\"effector_cluster_{idx}.mafft.fasta\", \"-o\", f\"effector_cluster_{idx}.mafft.hmm\", \"--name\", f\"effector_cluster_{idx}\", \"--acc\", f\"effector_cluster_{idx}\", \"--desc\", f\"effector_cluster_{idx}\"])\n",
    "!cat effector_cluster_*.mafft.hmm > effector_clusters.hmm\n",
    "\n",
    "!hmmer_compare.py -i effector_clusters.hmm -r ../resources/Pfam-A.hmm -o effector_clusters_vs_pfam.tsv --cpu 8 -k 10\n",
    "    \n",
    ";"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create table of similarity between clusters and Makarova et al. 2020 effectors\n",
    "!hmmer_compare.py -i effector_clusters.hmm -r effector_no_consensus/effectors.hmm -o effector_clusters_vs_makarova_effectors.tsv --cpu 8 -k 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pull random representatives from each cluster, to run AlphaFold2 on and do further annotation.\n",
    "import random\n",
    "seed  = 15\n",
    "seeded_random = random.Random(seed)\n",
    "\n",
    "with open(\"cluster_representatives.fasta\", \"w\") as out_handle:\n",
    "    for i in range(0, 12):\n",
    "        cluster = i + 1\n",
    "        with open(f\"effector_cluster_{cluster}.fasta\", \"r\") as in_handle:\n",
    "            seqs = [(title, seq) for title,seq in  SimpleFastaParser(in_handle)]\n",
    "            seeded_random.shuffle(seqs)\n",
    "            out_handle.write(f\">Cluster{cluster}_{seqs[0][0]}\\n{seqs[0][1]}\\n\")\n",
    "            out_handle.write(f\">Cluster{cluster}_{seqs[1][0]}\\n{seqs[1][1]}\\n\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# supplementary table generation\n",
    "\n",
    "! enum_report.py -i new_annots.gb  --length --domains --domain_descriptions -o annotations.Pfam.tsv --sequence --databases Pfam-A  --column_names \"Name\" \"Length\" \"Pfam 36 Domains\" \"Pfam 36 Domain Descriptions\" \"Sequence\"\n",
    "! enum_report.py -i new_annots.gb  --domains -o annotations.effector.tsv --databases effectors --column_names \"Name\" \"CARF effector Domains (Makarova et al. 2020)\"\n",
    "#! enum_report.py -i new_annots.gb --domains --domain_descriptions -o annotations.Pfam.tsv --length --sequence --databases Pfam-A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "pfam_annots = pd.read_table(\"annotations.Pfam.tsv\", sep=\"\\t\")\n",
    "effector_annots = pd.read_table(\"annotations.effector.tsv\", sep=\"\\t\")\n",
    "Pfam_with_effector = pd.merge(pfam_annots, effector_annots, on=\"Name\", how=\"outer\")\n",
    "clusters_table = pd.read_table(\"clusters.tsv\", names=[\"Name\", \"cluster\"])\n",
    "Pfam_with_effector_and_clusters = pd.merge(Pfam_with_effector, clusters_table, on=\"Name\", how=\"outer\")\n",
    "CARF_supplementary_table_2 = pd.merge(Pfam_with_effector_and_clusters, metadata[[\"Local ID\",\"Other Protein ID\", \"Final group assignments\"]], left_on=\"Name\", right_on=\"Local ID\", how=\"left\")\n",
    "\n",
    "def count_join(list_of_strings:list, input_sep = \"; \", output_sep = \", \"):\n",
    "    out = dict() #counts\n",
    "    for s in list_of_strings:\n",
    "        if s == \"\":\n",
    "            continue\n",
    "        parts = s.split(input_sep)\n",
    "        for part in parts:\n",
    "            if part in out:\n",
    "                out[part] += 1\n",
    "            else:\n",
    "                out[part] = 1\n",
    "    out_tuple = [(k, v) for k, v in out.items()]\n",
    "    out_tuple.sort(key=lambda x: x[1], reverse=True)\n",
    "    return output_sep.join([f\"{k} ({v})\" for k, v in out_tuple])\n",
    "\n",
    "\n",
    "# count name, join with comma Pfam 36 domains and CARF effector domains (Makarova et al. 2020), skip blank lines so that there aren't a bunch of commas at the end of the line\n",
    "# \n",
    "CARF_supplementary_table_2.fillna(\"\", inplace=True)\n",
    "#CARF_supplementary_table_1 = CARF_supplementary_table_2[[\"cluster\", \"Pfam 36 Domains\", \"CARF effector Domains (Makarova et al. 2020)\", \"Name\"]].groupby(\"cluster\").agg({\"Name\": \"count\", \"Pfam 36 Domains\": lambda x: \", \".join(x), \"CARF effector Domains (Makarova et al. 2020)\": lambda x: \", \".join(x)}).reset_index()\n",
    "CARF_supplementary_table_1 = CARF_supplementary_table_2[[\"cluster\", \"Pfam 36 Domains\", \"CARF effector Domains (Makarova et al. 2020)\", \"Final group assignments\", \"Name\"]].groupby(\"cluster\").agg({\"Name\": \"count\", \"Final group assignments\": lambda x: count_join(x), \"Pfam 36 Domains\": lambda x: count_join(x), \"CARF effector Domains (Makarova et al. 2020)\": lambda x: count_join(x)}).reset_index()\n",
    "# rename \"Name to \"count\"\n",
    "CARF_supplementary_table_1.rename(columns={\"Name\": \"count\", \"Final group assignments\": \"CARF group\"}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Name', 'Length', 'Pfam 36 Domains', 'Pfam 36 Domain Descriptions',\n",
       "       'Sequence', 'CARF effector Domains (Makarova et al. 2020)', 'cluster',\n",
       "       'Local ID', 'Other Protein ID', 'Final group assignments'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CARF_supplementary_table_2.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "CARF_supplementary_table_1.to_csv(\"CARF_supplementary_table_1.tsv\", sep=\"\\t\", index=False)\n",
    "CARF_supplementary_table_2.to_csv(\"CARF_supplementary_table_2.tsv\", sep=\"\\t\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "domainator",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
